File: frc2018_3/src/main/java/ca/warp7/frc2018_3/subsystems/Climber.java
package ca.warp7.frc2018_3.subsystems;

import ca.warp7.frc.core.ISubsystem;
import ca.warp7.frc.core.Robot;
import ca.warp7.frc.core.StateType;
import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
import edu.wpi.first.wpilibj.SpeedController;

import static ca.warp7.frc.Functions.limit;
import static ca.warp7.frc2018_3.Constants.kClimberPins;

public class Climber implements ISubsystem {

    private static final double kAbsoluteMaxOutputPower = 1.0;

    @InputField
    private final InputState mInputState = new InputState();
    @StateField
    private final CurrentState mCurrentState = new CurrentState();

    private SpeedController mClimberMotors;

    public void setSpeed(double speed) {
        mInputState.demandedSpeed = speed;
    }

    @Override
    public void onConstruct() {
        mClimberMotors = new WPI_VictorSPX(kClimberPins);
    }

    @Override
    public void onDisabled() {
        mInputState.demandedSpeed = 0;
    }

    @Override
    public void onOutput() {
        mClimberMotors.set(limit(mCurrentState.speed, kAbsoluteMaxOutputPower));
    }

    @Override
    public void onUpdateState() {
        mCurrentState.speed += (mInputState.demandedSpeed - mCurrentState.speed) / 6.0;
    }

    @Override
    public void onReportState() {
        Robot.report(this, StateType.ComponentInput, mInputState);
        Robot.report(this, StateType.ComponentState, mCurrentState);
    }

    static class InputState {
        double demandedSpeed;
    }

    static class CurrentState {
        double speed;
    }
}

File: frc2018_5/src/main/kotlin/ca/warp7/frc2018/v5/subsystems/Climber.kt
package ca.warp7.frc2018.v5.subsystems

import ca.warp7.frc.Subsystem
import ca.warp7.frc2018.v5.constants.RobotPins.kClimberPin
import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.can.VictorSPX

object Climber : Subsystem {
    private val climberMaster = VictorSPX(kClimberPin)

    var percentOutput = 0.0

    @Synchronized
    override fun onDisabled() = climberMaster.set(ControlMode.Disabled, 0.0)

    override fun onIdle() = onDisabled()

    @Synchronized
    override fun onOutput() = climberMaster.set(ControlMode.PercentOutput, percentOutput)
}
